FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageJobQueueClient.cs

```diff
@@ -0,0 +1,717 @@
+        }
+
+        // The expected behaviors:
+        // 1. multi agent instances enqueue same jobs concurrently, only one job entity will be created,
+        //    there may be multi messages, while only one will be recorded in job lock entity, and the created jobInfo will be returned for all instances.
+        // 2. re-enqueue job, no matter what the job status is now, will do nothing, and return the existing jobInfo.
+        // 3. if one of the steps fails, will continue to process it when re-enqueue.
+        // TODO: The parameter forceOneActiveJobGroup and isCompleted are ignored for now
+        public async Task<IEnumerable<JobInfo>> EnqueueAsync(
+            byte queueType,
+            string[] definitions,
+            long? groupId,
+            bool forceOneActiveJobGroup,
+            bool isCompleted,
+            CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Start to enqueue {definitions.Length} jobs.");
+
+            // step 1: get incremental job ids, will try again if fails
+            var jobIds = await GetIncrementalJobIds(queueType, definitions.Length, cancellationToken);
+
+            // step 2: generate job info entities and job lock entities batch
+            var jobInfos = definitions.Select((definition, i) => new TJobInfo
+                {
+                    Id = jobIds[i],
+                    QueueType = queueType,
+                    Status = JobStatus.Created,
+                    GroupId = groupId ?? 0,
+                    Definition = definition,
+                    Result = string.Empty,
+                    CancelRequested = false,
+                    CreateDate = DateTime.UtcNow,
+                    HeartbeatDateTime = DateTime.UtcNow,
+                })
+                .ToList();
+
+            var jobInfoEntities = jobInfos.Select(jobInfo => jobInfo.ToTableEntity()).ToList();
+            var jobLockEntities = jobInfoEntities.Select((jobInfoEntity, i) => new JobLockEntity
+                {
+                    PartitionKey = jobInfoEntity.PartitionKey,
+                    RowKey = AzureStorageKeyProvider.JobLockRowKey(jobInfos[i].JobIdentifier()),
+                    JobInfoEntityRowKey = jobInfoEntity.RowKey,
+                }).ToList();
+
+            // step 3: insert jobInfo entity and job lock entity in one transaction.
+            IEnumerable<TableTransactionAction> transactionActions = jobInfoEntities
+                .Select(entity => new TableTransactionAction(TableTransactionActionType.Add, entity))
+                .Concat(jobLockEntities.Select(entity => new TableTransactionAction(TableTransactionActionType.Add, entity)));
+            try
+            {
+                await _azureJobInfoTableClient.SubmitTransactionAsync(transactionActions, cancellationToken);
+            }
+            catch (RequestFailedException ex) when (IsSpecifiedErrorCode(ex, AzureStorageErrorCode.AddEntityAlreadyExistsErrorCode))
```

Comment:
Why not just throw exception here.

Reply:
Removed the check for jobLockEntity count